[Tree-widget]: Widget header + visual fixes

MartynasStrazdas opened the pull request:
Changes:
 - Added a search box based on mockups to the tree selector component
 - Renamed tree selector to widget header for clarity. (Internal)
 - Fixed Tree header button styling
 - Updated hierarchies react package and added buttons

grigasp commented:
Purely based on the attached demo:

- Missing spacing between Select / Input and Search button:
  ![image](https://github.com/user-attachments/assets/081afe73-2582-49f2-a392-62f82c365433)
- Input outline doesn't fit.
- Spacing above / below action buttons is inconsistent.
- The spinner during search progress is temporary, until we get skeleton, right?

MartynasStrazdas commented:
> Purely based on the attached demo:
> 
> * Missing spacing between Select / Input and Search button:
>   ![image](https://private-user-images.githubusercontent.com/35135765/412414578-081afe73-2582-49f2-a392-62f82c365433.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzNjQyOTQsIm5iZiI6MTczOTM2Mzk5NCwicGF0aCI6Ii8zNTEzNTc2NS80MTI0MTQ1NzgtMDgxYWZlNzMtMjU4Mi00OWYyLWEzOTItNjJmODJjMzY1NDMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEyVDEyMzk1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczMmZhODA0ZTUwNzUzMTFhN2ZiZjJlMTQzNmRiOWJjOTFiMjc5NGNiYjQ2ZDBkNTk4ZGIxMmJhMzcwZmNhNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mwvN_SVi4AIQniDBeLd7wBAkkF-smTovH-PcMtSTVl0)
> * Input outline doesn't fit.
> * Spacing above / below action buttons is inconsistent.
> * Are we using correct variables for spacing overall? It looks like mockups have more padding.
> * The spinner during search progress is temporary, until we get skeleton, right?

- Added spacing between select/search to icon button.

- For the input outline on focus it seems to be an itwinUI issue will make a demo for them later.

- Fixed the spacing for buttons.

- Right now correct paddings should be used, I'm not sure for what padding is used in mockups for search/button, using standard 8px (0.5rem) right now (visually it seems the mockup is using 0.5 rem as well).

- Did not touch loading, it is temporary until we get linear loading, I think we discussed using skeleton only for initial load.

grigasp commented:
All stuff in the header should be disable while we load available trees. Now it's usable:
![image](https://github.com/user-attachments/assets/095c9823-8be2-4a61-a28f-b2e3ec506dfd)

Also, the "search" tooltip string should start with upper case.

grigasp commented:
> * Added spacing between select/search to icon button.

👍 

> * For the input outline on focus it seems to be an itwinUI issue will make a demo for them later.

@GerardasB @mayank99 The outline is a problem on the buttons too:
![image](https://github.com/user-attachments/assets/e897a0fe-d241-4f31-b387-03dab5bc0407)
![image](https://github.com/user-attachments/assets/48aaa614-33c7-4549-a536-2545bb3efdd7)
![image](https://github.com/user-attachments/assets/1fc27a60-107f-47cc-a4d8-c124653e2795)

> * Fixed the spacing for buttons.

👍 

> * Right now correct paddings should be used, I'm not sure for what padding is used in mockups for search/button, using standard 8px (0.5rem) right now (visually it seems the mockup is using 0.5 rem as well).

As far as I can tell, the header in mockups is set to use 12px:
![image](https://github.com/user-attachments/assets/b78de0fa-23e6-406b-9056-7740960e64af)

> * Did not touch loading, it is temporary until we get linear loading, I think we discussed using skeleton only for initial load.

👍

grigasp commented:
**Dark mode**

![image](https://github.com/user-attachments/assets/6daeeeea-6bd9-40bf-b3a9-3790bc880d60)

- Our tree selector has a shadow. Mockups don't.
- The border / separator under header buttons seems to be of wrong color.
- Why is the tree background darker than the rest of the widget?

MartynasStrazdas commented:
> **Dark mode**
> 
> ![image](https://private-user-images.githubusercontent.com/35135765/412773868-6daeeeea-6bd9-40bf-b3a9-3790bc880d60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0MzIyMjIsIm5iZiI6MTczOTQzMTkyMiwicGF0aCI6Ii8zNTEzNTc2NS80MTI3NzM4NjgtNmRhZWVlZWEtNmJkOS00MGJmLWIzYTktMzc5MGJjODgwZDYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDA3MzIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5ZGI2MDQ0ZTNiYzExNjBmOWUyMDJhMDhiZjk5ZDJmZGY0ODdmYTViNjczOTcyOTAxZDY4ZWEzMzZhYTVmODEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zo-4_YFecYsr627qW8ctyHBuR2VwRxCRa6ELo63SnwU)
> 
> * Our tree selector has a shadow. Mockups don't.
> * The border / separator under header buttons seems to be of wrong color.
> * Why is the tree background darker than the rest of the widget?

![image](https://github.com/user-attachments/assets/23079e37-6e68-4ae9-9d7c-cce8ced77d89)
![image](https://github.com/user-attachments/assets/e8ef62ef-d69e-4bdf-ad98-84b241b4fbf5)

* Our tree selector has a shadow. Mockups don't. & Why is the tree background darker than the rest of the widget?
It seems the dark mode itself has incorrect background color, this could be because of theme bridge.
Light mode has tree have a gray background which should not be the case, kiwi sandbox does not seem to have this issue, still looking into it, but the fix for this will probably be in `hierarchies-react`.

 * The border / separator under header buttons seems to be of wrong color.
 Added an actual Seperator component, it is less visible in dark mode, because of incorrect background color.

grigasp commented:
>   Added an actual Seperator component, it is less visible in dark mode, because of incorrect background color.

Looks better, thanks. Why is there are space between the separator and the first node?

grigasp commented:
Also, while playing around with search and instances focus toggle button. Currently it gets disabled when we start searching (enter something in the input box). I think it would be better if it got disabled as soon as you click on the search button and the search input shows up.

MartynasStrazdas commented:
> Also, while playing around with search and instances focus toggle button. Currently it gets disabled when we start searching (enter something in the input box). I think it would be better if it got disabled as soon as you click on the search button and the search input shows up.

Wouldn't this cause unnecessary reloads/filtering, you click on focus mode it filters tree to selected items, you click on search it starts to reload tree to all nodes, and then you search it filters again.

MartynasStrazdas commented:
> > It seems the dark mode itself has incorrect background color, this could be because of theme bridge.
> > Light mode has tree have a gray background which should not be the case, kiwi sandbox does not seem to have this issue, still looking into it, but the fix for this will probably be in hierarchies-react.
> 
> This doesn't answer the question regarding shadow.

The shadow comes from kiwi side, will file an issue for it.

> > Added an actual Seperator component, it is less visible in dark mode, because of incorrect background color.
> 
> Looks better, thanks. Why is there are space between the separator and the first node?

In the mockups there is an additional padding after Seperator.

> > Ps. Adjusted paddings to reflect mockups better, this fixed the focus on selector/search problem on our end.
> 
> I think there's still a problem with outline on header buttons - the outline is displayed on top of sibling buttons. I think it's because we don't have any spacing between them.

This could be slightly tricky, because we allow buttons to be provided therefor, we can't add margins to the buttons itself. We could have something like this `.container-name button { margin-right: var(4px equivalent)} @saskliutas I'm not sure if this would be an appropriate solution as it could cause issues in the future. It would be best if kiwi would apply this focus border inside the component instead of outside.

grigasp commented:
> > Also, while playing around with search and instances focus toggle button. Currently it gets disabled when we start searching (enter something in the input box). I think it would be better if it got disabled as soon as you click on the search button and the search input shows up.
> 
> Wouldn't this cause unnecessary reloads/filtering, you click on focus mode it filters tree to selected items, you click on search it starts to reload tree to all nodes, and then you search it filters again.

Enabling focus mode and clicking on an element shows a filtered-down hierarchy. When I open the search, I expect it to be applied on the currently visible hierarchy, and that is not what's happening - the hierarchy changes to full as soon as I start typing. I think changing it to full as soon as user clicks on the search button makes it more obvious why that is happening and what the search is going to be applied on. I don't think we should worry too much about extra reloads/filtering.

grigasp commented:
> In the mockups there is an additional padding after Seperator.

In most - yes, in the one I was looking at - no 🤔 I'll ask UX for clarification...

saskliutas commented:
> > > It seems the dark mode itself has incorrect background color, this could be because of theme bridge.
> > > Light mode has tree have a gray background which should not be the case, kiwi sandbox does not seem to have this issue, still looking into it, but the fix for this will probably be in hierarchies-react.
> > 
> > 
> > This doesn't answer the question regarding shadow.
> 
> ~The shadow comes from kiwi side, will file an issue for it.~ Will fix
> 
> > > Added an actual Seperator component, it is less visible in dark mode, because of incorrect background color.
> > 
> > 
> > Looks better, thanks. Why is there are space between the separator and the first node?
> 
> In the mockups there is an additional padding after Seperator.
> 
> > > Ps. Adjusted paddings to reflect mockups better, this fixed the focus on selector/search problem on our end.
> > 
> > 
> > I think there's still a problem with outline on header buttons - the outline is displayed on top of sibling buttons. I think it's because we don't have any spacing between them.
> 
> This could be slightly tricky, because we allow buttons to be provided therefor, we can't add margins to the buttons itself. We could have something like this `.container-name button { margin-right: var(4px equivalent)} @saskliutas I'm not sure if this would be an appropriate solution as it could cause issues in the future. It would be best if kiwi would apply this focus border inside the component instead of outside.

I think ideally gaps between buttons would be handled by ButtonGroup component. For now you can put buttons in a flex container and set gap.

MartynasStrazdas commented:
Changed the focus mode do be disabled on search open. Added a gap for the buttons. For the select shadow waiting for a response from kiwi as the "outline" variant is currently broken in dark mode. 

https://github.com/user-attachments/assets/26163dfa-7459-42df-815c-77ca65cc3397

Ps. will have loading fixed in a separate pr

saskliutas commented on packages/itwin/tree-widget/src/tree-widget-react/components/tree-header/TreeHeader.tsx:
```diff
@@ -5,67 +5,33 @@
-}
-
-function HeaderButtons(props: PropsWithChildren<HeaderButtonsProps>) {
+function HeaderButtons(props: PropsWithChildren) {
```
Looks like redundant

saskliutas commented on packages/itwin/tree-widget/src/tree-widget-react/components/tree-header/TreeHeader.tsx:
```diff
@@ -5,67 +5,33 @@
-      </HeaderButtons>
-    </div>
+    <>
+      <div className={classnames("tw-tree-header", className)}>
```
Maybe this should be renamed to something like `tw-tree-tools`/`tw-tree-toolbar`. I also think it would be nice to avoid calling it TreeHeader as it might cause confusion with TreeWidgetHeader. Maybe something like `TreeToolbar`? @grigasp

grigasp replied:
agree, toolbar sounds better

MartynasStrazdas replied:
Changed to toolbar

saskliutas commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/TreeRenderer.tsx:
```diff
@@ -25,7 +25,10 @@ export function TreeRenderer({ rootNodes, onNodeClick, expandNode, visibilityBut
         rootNodes={rootNodes}
-        actionsRenderer={visibilityButtonProps ? (node) => <TreeItemVisibilityButton {...visibilityButtonProps} node={node} /> : undefined}
+        actions={[
+          ...(visibilityButtonProps ? [createVisibilityAction(visibilityButtonProps)] : []),
```
Can we move code related to visibility buttons into `VisibilityTreeRenderer`?

MartynasStrazdas replied:
The only issue would be that filter button would go first then visibility. If this order is fine, have the changes ready.

saskliutas replied:
I think we could duplicate filter creation in `VisibilityTreeRenderer`. This might result in use having 3 components: `TreeRenderer`, `VisibilityTreeRenderer` and `BaseTreeRenderer` (internal) but would allow to keep separate API for visibility and non-visibility trees. It could look like:

```tsx
function BaseTreeRenderer(props) {
  // setup and render tree
  return <TreeComp
    actions={props.actions}
  />
}

function TreeRenderer(props) {
  return <BaseTreeRenderer 
    actions={]createFillterAction(), ...props.consumerActions]}
  />;
}

function VisibilityTreeRenderer(props) {
  return <BaseTreeRenderer 
    actions={[createVisibilityAction(), createFilterAction(), ...props.consumerActions]}
  />
}
```

MartynasStrazdas replied:
Changed and added a way to provide actions to all base trees, which was missing.

saskliutas reviewed (changes requested):
I saw that you added `.css` file instead of `.scss`. I think we should be consistent and stick to one format. We do not have much css so I am fine with switching to `.css` instead of `.scss`.

saskliutas commented on packages/itwin/tree-widget/api/tree-widget-react.api.md:
```diff
@@ -33,6 +33,12 @@ import type { useSelectionHandler } from '@itwin/presentation-hierarchies-react'
 import type { Widget } from '@itwin/appui-react';
 
+// @internal (undocumented)
+function BaseTreeRenderer({ rootNodes, onNodeClick, expandNode, actions, ...props }: BaseTreeRendererProps): JSX_2.Element;
```
These should not be exported.

saskliutas commented on packages/itwin/tree-widget/api/tree-widget-react.api.md:
```diff
@@ -421,27 +389,48 @@ export class TreeWidget {
 
 // @public
-export function TreeWidgetComponent(props: SelectableTreeProps): JSX_2.Element;
+export function TreeWidgetComponent(props: TreeWidgetComponentProps): JSX_2.Element | null;
```
I reallyl don't like the naming with all those `TreeWidgetSomething`. Maybe we could do something like:
`TreeWidgetComponent` -> `TreeSelector` (probably could make it internal)
`TreeWidgetComponentDefinition` -> `TreeDefinition`
`TreeWidgetComponentProvider` -> `TreeWidgetComponent` (revert to the previous name)
